bee-message/src/payload/transaction/essence/regular.rs

```diff
@@ -0,0 +1,250 @@
+            return Err(Error::InvalidInputOutputCount(self.inputs.len()));
+        }
+
+        if !INPUT_OUTPUT_COUNT_RANGE.contains(&self.outputs.len()) {
+            return Err(Error::InvalidInputOutputCount(self.outputs.len()));
+        }
+
+        if !matches!(self.payload, None | Some(Payload::Indexation(_))) {
+            // Unwrap is fine because we just checked that the Option is not None.
+            return Err(Error::InvalidPayloadKind(self.payload.unwrap().kind()));
+        }
+
+        // Inputs validation
+
+        for input in self.inputs.iter() {
+            match input {
+                Input::Utxo(_) => {
+                    if self.inputs.iter().filter(|i| *i == input).count() > 1 {
+                        return Err(Error::DuplicateError);
+                    }
+                }
+                _ => return Err(Error::InvalidInputKind(input.kind())),
+            }
+        }
+
+        // Inputs must be lexicographically sorted in their serialised forms.
+        if !is_sorted(self.inputs.iter().map(Packable::pack_new)) {
+            return Err(Error::TransactionInputsNotSorted);
+        }
+
+        // Outputs validation
+
+        let mut total: u64 = 0;
+
+        // TODO iteration-based or memory-based ?
+
+        for output in self.outputs.iter() {
+            match output {
+                Output::SignatureLockedSingle(single) => {
+                    // The addresses must be unique in the set of SignatureLockedSingleOutputs.
+                    if self
+                        .outputs
+                        .iter()
+                        .filter(|o| matches!(o, Output::SignatureLockedSingle(s) if s.address() == single.address()))
+                        .count()
+                        > 1
+                    {
+                        return Err(Error::DuplicateError);
+                    }
+
+                    total = total
+                        .checked_add(single.amount())
+                        .ok_or_else(|| Error::InvalidAccumulatedOutput((total + single.amount()) as u128))?;
+                }
+                Output::SignatureLockedDustAllowance(dust_allowance) => {
+                    // The addresses must be unique in the set of SignatureLockedDustAllowanceOutputs.
+                    if self
+                        .outputs
+                        .iter()
+                        .filter(
+                            |o| matches!(o, Output::SignatureLockedDustAllowance(s) if s.address() == dust_allowance.address()),
+                        )
+                        .count()
+                        > 1
+                    {
+                        return Err(Error::DuplicateError);
+                    }
+
+                    total = total
+                        .checked_add(dust_allowance.amount())
+                        .ok_or_else(|| Error::InvalidAccumulatedOutput((total + dust_allowance.amount()) as u128))?;
```

Comment:
Creating the error will potentially panic or cause UB at runtime if that overflows. You need to cast each summand `as u128`, not the already overflowed sum.

Reply:
It would only panic on debug mode but you're right, the value reported would be weird, I'll change

Reply:
Nice catch, thanks